etl_pipeline/transform.py

```diff
@@ -0,0 +1,210 @@
+    return 4
+
+
+def get_weather_alerts(weather_data: pd.DataFrame) -> pd.DataFrame:
+    """Obtain a dataframe of alerts for a forecast."""
+    alerts = pd.DataFrame()
+    twelve_hour_data = weather_data[weather_data['forecast_timestamp'] < datetime.now(
+    ) + timedelta(hours=12)]
+    alerts['heat_warning'] = twelve_hour_data['temperature'].apply(
+        calculate_heat_alerts)
+    alerts['wind_warning'] = twelve_hour_data[['wind_gusts', 'wind_speed']].apply(
+        lambda x: calculate_wind_alerts(x['wind_gusts'], x['wind_speed']), axis=1)
+    alerts['ice_warning'] = twelve_hour_data['temperature'].apply(
+        calculate_ice_alerts)
+    alerts['lightning_warning'] = twelve_hour_data['lightning_potential'].apply(
+        calculate_lightning_alerts)
+    alerts['snowfall_warning'] = twelve_hour_data['snowfall'].apply(
+        calculate_snowfall_alerts)
+    alerts['visibility_warning'] = twelve_hour_data['visibility'].apply(
+        calculate_visibility_alerts)
+    alerts['uv_warning'] = twelve_hour_data['uv_index'].apply(
+        calculate_uv_alerts)
+    alerts['rainfall_warning'] = twelve_hour_data['rainfall'].apply(
+        calculate_rain_alerts)
+    return alerts
+
+
+def gather_data_from_json(json_data: dict, key: str) -> list[dict]:
+    """Obtain transformed forecast data from hourly or 15-minutely data."""
+    data = pd.DataFrame(json_data[key])
+    data = data.map(lambda x: 0 if x is None or x == 'null' else x)
+    data = rename_columns(data)
+    data = change_data_types(data)
+    alerts = get_weather_alerts(data)
+    forecasts = data.to_dict(orient="records")
+    warnings = alerts.itertuples(index=False, name=None)
+    forecast_warnings = (create_warning_list(forecast)
+                         for forecast in warnings)
+    return list({'forecast': x, 'warnings': y} for x, y
+                in zip_longest(forecasts, forecast_warnings))
+
+
+def gather_weather_data(latitude: float, longitude: float) -> list[dict]:
+    """Obtain transformed forecast data for the whole weather report."""
+    minutely_data = get_weather_details_for_24hrs(latitude, longitude)
+    hourly_data = get_weather_details_for_week(latitude, longitude)
+    return gather_data_from_json(minutely_data, 'minutely_15') +\
+        gather_data_from_json(hourly_data, 'hourly')
+
+
+def gather_air_quality(latitude: float, longitude: float) -> dict:
+    """Obtain the o3 air quality for the weather report."""
+    concentration = get_air_quality(latitude, longitude)['O3']['concentration']
+    try:
+        concentration = float(concentration)
+    except ValueError:
+        concentration = 'error'
+    if concentration < 0:
+        concentration = 'error'
+    severity_id = calculate_air_quality_alert(concentration)
+    return {'o3_concentration': concentration, 'severity_id': severity_id}
+
+
+if __name__ == "__main__":
+    load_dotenv()
+    for x in gather_weather_data(0, 0):
+        print(x)
+    print(gather_air_quality(0, 0))
```

Comment:
do we need to print these or just a debugging artifact?

Reply:
removed